src/services/refactors/moveToFile.ts

```diff
@@ -0,0 +1,198 @@
+            return [{ name: refactorNameForMoveToFile, description, actions:
+                [{ ...moveToFileAction, notApplicableReason: getLocaleSpecificMessage(Diagnostics.Selection_is_not_a_valid_statement_or_statements) }]
+            }];
+        }
+        return emptyArray;
+    },
+    getEditsForAction: function getRefactorEditsToMoveToFile(context, actionName, newFile): RefactorEditInfo | undefined {
+        Debug.assert(actionName === refactorNameForMoveToFile, "Wrong refactor invoked");
+        const statements = Debug.checkDefined(getStatementsToMove(context));
+        if (newFile) {
+            const edits = textChanges.ChangeTracker.with(context, t => doChange(context, context.file, newFile, context.program, statements, t, context.host, context.preferences));
+            return { edits, renameFilename: undefined, renameLocation: undefined };
+        }
+        return undefined;
+    }
+});
+
+function doChange(context: RefactorContext, oldFile: SourceFile, newFile: string, program: Program, toMove: ToMove, changes: textChanges.ChangeTracker, host: LanguageServiceHost, preferences: UserPreferences): void {
+    const checker = program.getTypeChecker();
+    const usage = getUsageInfo(oldFile, toMove.all, checker);
+    //creating a new file
+    if (!host.fileExists(newFile) || host.fileExists(newFile) && program.getSourceFile(newFile)?.statements.length === 0) {
+        changes.createNewFile(oldFile, newFile, getNewStatementsAndRemoveFromOldFile(oldFile, newFile, usage, changes, toMove, program, host, newFile, preferences, /*newFileExists*/ false));
+        addNewFileToTsconfig(program, changes, oldFile.fileName, newFile, hostGetCanonicalFileName(host));
+    }
+    else {
+        const sourceFile = program.getSourceFile(newFile);
+        if (sourceFile) {
+            const importAdder = codefix.createImportAdder(sourceFile, context.program, context.preferences, context.host);
+            getNewStatementsAndRemoveFromOldFile(oldFile, newFile, usage, changes, toMove, program, host, newFile, preferences, /*newFileExists*/ true, importAdder);
+        }
+    }
+}
+
+function getNewStatementsAndRemoveFromOldFile(
+    oldFile: SourceFile, newFile: string, usage: UsageInfo, changes: textChanges.ChangeTracker, toMove: ToMove, program: Program, host: LanguageServiceHost, newFilename: string, preferences: UserPreferences, newFileExists: boolean, importAdder?: codefix.ImportAdder
+) {
+    const checker = program.getTypeChecker();
+    const prologueDirectives = takeWhile(oldFile.statements, isPrologueDirective);
+    if (oldFile.externalModuleIndicator === undefined && oldFile.commonJsModuleIndicator === undefined && usage.oldImportsNeededByNewFile.size === 0 && !newFileExists) {
+        deleteMovedStatements(oldFile, toMove.ranges, changes);
+        return [...prologueDirectives, ...toMove.all];
+    }
+
+    const useEsModuleSyntax = !!oldFile.externalModuleIndicator;

```

Comment:
This is a very outdated assumption now. We should generally assume that we should use ESM syntax unless we explicitly see CJS syntax. Also, what you get when this is `false` is _JavaScript_ CJS syntax that won’t work in TypeScript files. So the calculation of this variable needs to be quite a bit more sophisticated now. It needs to consider:

- Is `newFile` TypeScript or JavaScript?
- What is the `impliedNodeFormat` of the new file going to be?
- Is `verbatimModuleSyntax` enabled?

I think the value `useEsModuleSyntax` should only be `false` when `newFile` is JS, the new file’s `impliedNodeFormat` is not `ESNext`, and either `oldFile.commonJsModuleIndicator` is truthy or `verbatimModuleSyntax` is enabled.

Reply:
The simplest relevant test to add would be one where the file you’re moving from has no imports or exports.